### ISSUE TITLE
Any way to generate an HTTP 403 upon drop?

### ISSUE THREAD

**Issue by WhiteWinterWolf** [jvoisin/snuffleupagus#385]
Currently a request dropped by Snuffleupagus results in a WSOD.

Is there any way to generate to generate a HTTP 403 error message instead?

This would be both:

- **More informative for legitimate users**, as they will be informed that their request has been intentionally blocked due to the detection of a suspicious content or behavior.
- **Less informative for attackers**, as the server would expose a consistent behavior no matter if the request has been blocked at the WAF or the Snuffleupagus layer.

**Comment by jvoisin** [comment]
There is currently no way to do this from Snuffleupagus' side, but this can easily be done in your web server or reverse proxy. For example, in Apache: `ErrorDocument 403 /custom_403_handler.php`.

I don't have plans to implement this directly in Snuffleupagus.

**Comment by WhiteWinterWolf** [comment]
Thanks for your quick answer :) .

My question is in fact on the step just earlier.

Currently, when a Snuffleupagus drops a request, the result is an empty `HTTP/1.1 200 OK` reply (classic WSOD), while I wonder if there is a way to produce a `HTTP1/1 403 Forbidden`. Only then I could customize the 403 error content to match whatever I want using the statement you indicate.

**Comment by wargio** [comment]
i do not think this is possible due the code calling the `zend_bailout` function

**Comment by WhiteWinterWolf** [comment]
Conceptually, I would say that this is wrong that a request *rejected* by Snuffleupagus generates an `HTTP 200` status which means that *"The request has succeeded"* ([RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.1)).

However, I suppose that on the dev side you are bound to some PHP.... "oddities", as can be seen in [this](https://bugs.php.net/bug.php?id=50921) dozen years long (!) thread, debating whether parsing errors should generate `HTTP 200` or `HTTP 500` status. While Snuffleupagus saves us from the most damaging of these "oddities" (and I warmly thank you about that :) ), it cannot save us from everything ;) !

If it is confirmed that there is no way for Snuffleupagus to send anything else than a `HTTP 200` status for rejected queries, is there maybe another data which could be injected somewhere (a header, a keyword in the content section, etc.) and would allow an upstream WAF to reliably catch those replies and react accordingly ?

**Comment by jvoisin** [comment]
An ugly way to do this would be to patch Snuffleupagus to insert calls to `header`, but I'm not convinced we should go this way.

I don't understand why you want to plug your WAF on Snuffleupagus. If it's for a fail2ban-like behaviour, you can simply watch its logs.

**Comment by WhiteWinterWolf** [comment]
If it is confirmed that this issue cannot be solved *inside* Snuffleupagus due to PHP limitations, then I'm thinking about potential workarounds *outside* of it.

I search a way to distinguish between queries rejected by Snuffleupagus' policy, queries aborted due to a fatal error server-side, and queries processed successfully. In an ideal world, this is precisely what HTTP status codes are made for: 403, 500 and 200 respectively. However, it appears that PHP stubbornly enforces a HTTP status 200 for everything, and as @wargio mentioned chances are this is out of reach of Snuffleupagus code.

The only way I can imagine for now is to closely compare the replies sent in those various scenarios and maybe find some kind of heuristic (a missing header for instance) telling that a given empty reply is most likely not a real `HTTP 200 Success` but Snuffleupagus dropping the request. This would allow to trigger something like the `custom_403_handler.php` you mentioned earlier. But relying on such mechanism would be seriously ugly and prone to false positives.

Hence my question : if it is confirmed that it is not possible for Snuffleupagus to set a proper HTTP status, would it possible to set something else which could be used to reliably distinguish a dropped query from a successful one? Or are we deemed to rely on ugly tricks to cope with those "`HTTP 200 Maybe-Success-Maybe-Not`" ?

As a side note, while Snuffleupagus logs are sufficient to investigate an issue, it cannot and is not meant (IMHO, but maybe I missed some advanced logging features) to replace full-fledged and flexible WAF logs (with logging of the remote IP instead of the reverse-proxy one, logging of a unique request ID, etc.).

**Comment by jvoisin** [comment]
It's possible if you patch it to send headers before calling `zend_bailout()`, but it's a bit gross :/

**Comment by WhiteWinterWolf** [comment]
That's why I opened it as question, wondering if there is any proper way to handle this, whether inside Snuffleupagus or outside of it (Apache or PHP configuration), but it seems there isn't so I think this issue can be closed.

I could open another one on PHP's side, but as I linked above there is already [a similar issue ](https://bugs.php.net/bug.php?id=50921) opened in 2010 which still opened and being "discussed" nowadays, 11 years later, so I don't think it will progress in any way.

Let's live with this!
